From 039d5b33c1626cb2e4b49233e87ad1819e9762ab Mon Sep 17 00:00:00 2001
From: Roman Janota <Roman.Janota@cesnet.cz>
Date: Fri, 2 Oct 2026 14:33:25 +0200
Subject: [PATCH] sn BUGFIX dampening modify under the timer lock

Modifying the dampening period read the timer thread ID and the pending
change notification without the dampening timer lock. Both are also
written by the on-change callback and the dampening timer, and since a
one-shot timer clears its ID once it fires, the timer could disarm
between the check and the update.

The period and the next due time are now set under the lock, the timer
is updated whether armed or not, and a period of 0 flushes the pending
notification under the lock once the timer is stopped.

Upstream: https://github.com/sysrepo/sysrepo/commit/1b8518e7f3e68bee975f8a8f5db4f6818e8ca4cb
Signed-off-by: Mattias Walström <lazzer@gmail.com>
---
 src/utils/sn_yang_push.c | 44 ++++++++++++++++++++++++++--------------
 1 file changed, 29 insertions(+), 15 deletions(-)

diff --git a/src/utils/sn_yang_push.c b/src/utils/sn_yang_push.c
index 83c5b1aa..a28aeea2 100644
--- a/src/utils/sn_yang_push.c
+++ b/src/utils/sn_yang_push.c
@@ -1005,25 +1005,39 @@ srsn_yp_on_change_modify(struct srsn_sub *sub, uint32_t dampening_period_ms)
     sr_error_info_t *err_info = NULL;
     struct timespec next_notif;
 
-    if (dampening_period_ms != sub->dampening_period_ms) {
-        if (!dampening_period_ms) {
-            /* stop the timer */
-            srsn_update_timer(NULL, NULL, &sub->damp_sntimer);
+    if (dampening_period_ms == sub->dampening_period_ms) {
+        goto cleanup;
+    }
 
-            /* send the prepared notification */
-            if (sub->change_ntf && (err_info = srsn_yp_ntf_change_send(sub))) {
-                goto cleanup;
-            }
-        } else if (sub->damp_sntimer.tid) {
-            /* learn when the next notification is due */
-            next_notif = sr_time_ts_add(&sub->last_notif, dampening_period_ms);
+    /* TIMER LOCK */
+    pthread_mutex_lock(&sub->damp_sntimer.lock);
+
+    /* set the new period before the timer is stopped */
+    sub->dampening_period_ms = dampening_period_ms;
+
+    /* learn when the next notification is due */
+    next_notif = sr_time_ts_add(&sub->last_notif, dampening_period_ms);
+
+    /* TIMER UNLOCK */
+    pthread_mutex_unlock(&sub->damp_sntimer.lock);
+
+    if (dampening_period_ms) {
+        /* move the timer of a dampened change, if there is one */
+        srsn_update_timer(&next_notif, NULL, &sub->damp_sntimer);
+    } else {
+        /* stop the timer */
+        srsn_update_timer(NULL, NULL, &sub->damp_sntimer);
+
+        /* TIMER LOCK */
+        pthread_mutex_lock(&sub->damp_sntimer.lock);
 
-            /* update the timer */
-            srsn_update_timer(&next_notif, NULL, &sub->damp_sntimer);
+        /* send the prepared notification */
+        if (sub->change_ntf) {
+            err_info = srsn_yp_ntf_change_send(sub);
         }
 
-        /* update stored params */
-        sub->dampening_period_ms = dampening_period_ms;
+        /* TIMER UNLOCK */
+        pthread_mutex_unlock(&sub->damp_sntimer.lock);
     }
 
 cleanup:
